From 7ebaa6d69d0abf885a7db1352bfa50a2b35c7dc5 Mon Sep 17 00:00:00 2001
From: Srujana Challa <schalla@marvell.com>
Date: Fri, 21 May 2021 18:53:50 +0530
Subject: [PATCH 1539/1921] crypto: cn10k: allocate 64KB for each microcode
 image

CN10K ASIM CPT eref is trying to fetch 64KB microcode data
all the time from given microcode image address but existing
CPT PF driver allocates the memory for actual microcode image
size. This is causing SMMU errors when 4K PAGE SIZE is used.
So, this patch changes the driver logic to allocate 64KB all
the time.

Signed-off-by: Srujana Challa <schalla@marvell.com>
Change-Id: I0fcb14588f296e2f706a33cad3f3acb565ddd05b
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/52378
Reviewed-by: Jerin Jacob Kollanukkaran <jerinj@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <sgoutham@marvell.com>
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/crypto/marvell/cn10k/cn10k_cptpf_ucode.c | 4 ++--
 drivers/crypto/marvell/cn10k/cn10k_cptpf_ucode.h | 2 ++
 2 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/drivers/crypto/marvell/cn10k/cn10k_cptpf_ucode.c b/drivers/crypto/marvell/cn10k/cn10k_cptpf_ucode.c
index 8e2f699e3b2e..1fe5b760e7e1 100644
--- a/drivers/crypto/marvell/cn10k/cn10k_cptpf_ucode.c
+++ b/drivers/crypto/marvell/cn10k/cn10k_cptpf_ucode.c
@@ -975,7 +975,7 @@ static int create_sysfs_eng_grps_info(struct device *dev,
 static void ucode_unload(struct device *dev, struct cn10k_cpt_ucode *ucode)
 {
 	if (ucode->va) {
-		dma_free_coherent(dev, ucode->size, ucode->va, ucode->dma);
+		dma_free_coherent(dev, CN10K_CPT_UCODE_SZ, ucode->va, ucode->dma);
 		ucode->va = NULL;
 		ucode->dma = 0;
 		ucode->size = 0;
@@ -994,7 +994,7 @@ static int copy_ucode_to_dma_mem(struct device *dev,
 	u32 i;
 
 	/*  Allocate DMAable space */
-	ucode->va = dma_alloc_coherent(dev, ucode->size, &ucode->dma,
+	ucode->va = dma_alloc_coherent(dev, CN10K_CPT_UCODE_SZ, &ucode->dma,
 				       GFP_KERNEL);
 	if (!ucode->va)
 		return -ENOMEM;
diff --git a/drivers/crypto/marvell/cn10k/cn10k_cptpf_ucode.h b/drivers/crypto/marvell/cn10k/cn10k_cptpf_ucode.h
index 6a3268d48446..4ebbe2cce6ab 100644
--- a/drivers/crypto/marvell/cn10k/cn10k_cptpf_ucode.h
+++ b/drivers/crypto/marvell/cn10k/cn10k_cptpf_ucode.h
@@ -31,6 +31,8 @@
 
 #define CN10K_CPT_ENGS_BITMASK_LEN   BITS_TO_LONGS(CN10K_CPT_MAX_ENGINES)
 
+#define CN10K_CPT_UCODE_SZ           (64 * 1024)
+
 /* Microcode types */
 enum cn10k_cpt_ucode_type {
 	CN10K_CPT_AE_UC_TYPE = 1,  /* AE-MAIN */
-- 
2.31.1

